### PR TITLE
fix: make sure to prefilter the flat portion of a combined knn

### LINKED ISSUE #1561: If there is new data, and a prefilter, then the new data is not being prefiltered
Given a dataset that has new data, and a nearest/vector query that has a prefilter, then the new data is not being subjected to the prefilter and may be incorrectly returned.  The query plan generated is:

```
Projection: fields=[vector, indexed, not_indexed, _distance]
  Take: columns="_distance, _rowid, vector, indexed, not_indexed"
    KNNFlat: k=10 metric=l2
      RepartitionExec: partitioning=RoundRobinBatch(1), input_partitions=2
        UnionExec
          Projection: fields=[_distance, _rowid, vector]
            KNNFlat: k=10 metric=l2 <--- THERE SHOULD BE A FILTER JUST BEFORE HERE
              LanceScan: uri=tmp/.tmpLth8FX/data, projection=[vector], row_id=true, ordered=false
          Take: columns="_distance, _rowid, vector"
            KNNIndex: name=6d1cb8b7-ec0a-4255-b0bd-c48c1b76c3c8, k=10
              FilterExec: indexed != 50
                LanceScan: uri=tmp/.tmpLth8FX/data, projection=[indexed], row_id=true, ordered=true
```

### LINKED ISSUE #1563: Empty prefiltering criteria results in exception
(from user report)

# Problem

Using prefiltering, if the filter results in an empty set, then an exception is raised.

# Expected behavior

Return an empty arrow table

# Stacktrace

```
File /usr/local/lib/python3.8/dist-packages/lance/dataset.py:308, in LanceDataset.to_table(self, columns, filter, limit, offset, nearest, batch_size, batch_readahead, fragment_readahead, scan_in_order, prefilter, with_row_id)
    247 def to_table(
    248     self,
    249     columns: Optional[list[str]] = None,
   (...)
    260     with_row_id: bool = False,
    261 ) -> pa.Table:
    262     """Read the data into memory as a pyarrow Table.
    263 
...
File /usr/local/lib/python3.8/dist-packages/pyarrow/ipc.pxi:754, in pyarrow.lib.RecordBatchReader.read_all()

File /usr/local/lib/python3.8/dist-packages/pyarrow/error.pxi:115, in pyarrow.lib.check_status()

OSError: Io error: Execution error: External error: Execution error: LanceError(IO): Execution error: Failed to compute distances: LanceError(Arrow): Schema error: Can not append column _distance on schema: Schema { fields: [Field { name: "_distance", data_type: Float32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "_rowid", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "image_intro", data_type: FixedSizeList(Field { name: "item", data_type: Float32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, 1280), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, /home/runner/work/lance/lance/rust/lance-index/src/vector/flat.rs:57:36, /home/runner/work/lance/lance/rust/lance/src/io/exec/take.rs:72:42
```

# Reproduction

user report:

"""
I reproduced with:
Do a normal index search, and the result looks good.
Add a filter that has obviously incorrect condition, such as start_time stamp is larger than end_time stamp, the error raises
"""

Confirmed that this was using prefilter=True. When prefilter=True is removed, it won't raise error

### PR DESCRIPTION
**Pull request by westonpace**
Closes #1561 
Closes #1563

### CODE REVIEW

File: rust/lance/src/datafusion/physical_expr.rs [hunk 1]
```diff
@@ -129,6 +129,13 @@ pub fn column_names_in_expr(expr: &dyn PhysicalExpr) -> Vec<String> {
     visitor.columns
 }
 
+pub fn unique_column_names_in_expr(expr: &dyn PhysicalExpr) -> Vec<String> {
+    let mut cols = column_names_in_expr(expr);
+    cols.sort();
+    cols.dedup();
+    cols
+}
```
Comment (westonpace) [inline rust/lance/src/datafusion/physical_expr.rs:137]: Alternatively, we could just make this the default behavior of `column_names_in_expr`.  thoughts @wjones127 ?

Comment (wjones127) [reply rust/lance/src/datafusion/physical_expr.rs:137]: In https://github.com/lancedb/lance/pull/1581, I change `column_names_in_expr` to return distinct columns already. I think that's what we always want. Let me know what you think.

Comment (westonpace) [reply rust/lance/src/datafusion/physical_expr.rs:137]: Yep, awesome.  I'll take a look later.

File: rust/lance-index/src/vector/flat.rs [hunk 2]
```diff
@@ -54,8 +54,12 @@ pub async fn flat_search(
...
 
     if batches.is_empty() {
-        let schema_with_distance = input_schema.try_with_column(distance_field())?;
-        return Ok(RecordBatch::new_empty(schema_with_distance.into()));
+        if input_schema.column_with_name(DIST_COL).is_none() {
+            let schema_with_distance = input_schema.try_with_column(distance_field())?;
+            return Ok(RecordBatch::new_empty(schema_with_distance.into()));
+        } else {
+            return Ok(RecordBatch::new_empty(input_schema));
+        }
```
Comment (westonpace) [inline rust/lance-index/src/vector/flat.rs:62]: This should also address #1563

File: rust/lance/src/dataset/scanner.rs [hunk 3]
```diff
@@ -884,16 +893,21 @@ impl Scanner {
...
                 false,
             );
+            if let Some(refine_expr) = filter_plan.refine_expr.as_ref() {
+                // If there is a prefilter we need to manually apply it to the new data
+                scan_node = Arc::new(FilterExec::try_new(refine_expr.clone(), scan_node)?);
+            }
             // first we do flat search on just the new data
             let topk_appended = self.flat_knn(scan_node, q)?;
 
+            dbg!(&topk_appended.schema());
```
Comment (wjones127) [inline rust/lance/src/dataset/scanner.rs:903]: Can this be removed?

**Review by wjones127 — CHANGES_REQUESTED** [review]
One thing we test is the plan structure, like here: https://github.com/lancedb/lance/blob/214dfa3be572131bd747ddaccd15cab39153ba77/rust/lance/src/dataset/scanner.rs#L1800

Perhaps we could make a test asserting there is a filter node as expected when doing combined KNN?

**Comment by westonpace** [comment]
> Perhaps we could make a test asserting there is a filter node as expected when doing combined KNN?

I went ahead and made a test focused on making sure that a filter is applied to the new data (prefilter or not).  It doesn't look for specific nodes though as I worry that might change with time.  It works by checking the results.  Let me know if you'd rather I put a test in looking for specific nodes.

**Review by wjones127 — APPROVED** [review]

**Pull request merged**